sdk/client-core/core/src/main/java/io/clientcore/core/util/SharedExecutorService.java

```diff
@@ -0,0 +1,371 @@
+import io.clientcore.core.implementation.util.EnvironmentConfiguration;
+import io.clientcore.core.implementation.util.ImplUtils;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+/**
+ * An {@link ExecutorService} that is shared by multiple consumers.
+ * <p>
+ * The shared executor service is created using the following configuration settings:
+ * <ul>
+ *     <li>{@code azure.sdk.shared.threadpool.maxpoolsize} system property or
+ *     {@code AZURE_SDK_SHARED_THREADPOOL_MAXPOOLSIZE} environment variable - The maximum pool size of the shared
+ *     executor service. If not set, it defaults to 10 times the number of available processors.</li>
+ *     <li>{@code azure.sdk.shared.threadpool.keepalivemillis} system property or
+ *     {code AZURE_SDK_SHARED_THREADPOOL_KEEPALIVEMILLIS} environment variable - The keep alive time in millis for
+ *     threads in the shared executor service. If not set, it defaults to 60 seconds. Limited to integer size.</li>
+ *     <li>{@code azure.sdk.shared.threadpool.usevirtualthreads} system property or
+ *     {@code AZURE_SDK_SHARED_THREADPOOL_USEVIRTUALTHREADS} environment variable - A boolean flag to indicate if the
+ *     shared executor service should use virtual threads. If not set, it defaults to true. Ignored if virtual threads
+ *     are not available in the runtime.</li>
+ * </ul>
+ */
+@SuppressWarnings({ "resource", "NullableProblems" })
+public final class SharedExecutorService implements ExecutorService {
+    private static final ClientLogger LOGGER = new ClientLogger(SharedExecutorService.class);
+
+    // Shared thread counter for all instances of SharedExecutorService created using the empty factory method.
+    private static final AtomicLong AZURE_SDK_THREAD_COUNTER = new AtomicLong();
+    private static final String AZURE_SDK_THREAD_NAME = "azure-sdk-global-thread-";
+
+    // The thread pool size for the shared executor service.
+    private static final int THREAD_POOL_SIZE;
+
+    // The thread pool keep alive time for the shared executor service.
+    private static final int THREAD_POOL_KEEP_ALIVE_MILLIS;
+
+    // Virtual thread support for the shared executor service.
+    private static final boolean THREAD_POOL_VIRTUAL;
+
+    private static final SharedExecutorService INSTANCE;
+
+    static {
+        THREAD_POOL_SIZE
+            = getConfig("azure.sdk.shared.threadpool.maxpoolsize", "AZURE_SDK_SHARED_THREADPOOL_MAXPOOLSIZE",
```

Comment:
We should not use `AZURE` or `azure` here. We should use `clientcore` instead.

Reply:
Good catch, will clean that up